### PR TITLE
add boltdb and portobuf

### PR DESCRIPTION
**Pull request by upccup**

### CODE REVIEW

**Comment by upccup** [comment]
[hold on]  need add dependency pkgs to vender

File: backend/create.go [hunk 1]
```diff
@@ -34,40 +34,35 @@ func (b *Backend) LaunchApplication(version *types.ApplicationVersion) error {
...
 						*taskInfo.Container.Docker.PortMappings[0].HostPort,
 						version.ID); err != nil {
 					}
 					for _, healthCheck := range task.HealthChecks {
 						check := types.Check{
 							ID:       task.Name,
-							Address:  *task.AgentHostname,
-							Port:     int(*taskInfo.Container.Docker.PortMappings[0].HostPort),
+							Address:  task.AgentHostname,
+							Port:     int64(*taskInfo.Container.Docker.PortMappings[0].HostPort),
```
Comment (cmingxu) [inline backend/create.go]: PortMappings 一定有数据么

File: backend/list.go [hunk 2]
```diff
@@ -5,14 +5,14 @@ import (
 )
 
 func (b *Backend) ListApplications() ([]*types.Application, error) {
-	return b.store.ListApplications()
+	return b.store.GetApps()
 }
 
 func (b *Backend) ListApplicationTasks(id string) ([]*types.Task, error) {
-	return b.store.ListApplicationTasks(id)
+	return b.store.GetTasks(id)
```
Comment (cmingxu) [inline backend/list.go]: why GetTasks not GetTask

File: backend/list.go [hunk 3]
```diff
@@ -5,14 +5,14 @@ import (
...
 }
 
 // ListApplicationVersions is used to list all versions for application from consul specified by application id.
-func (b *Backend) ListApplicationVersions(applicationId string) ([]string, error) {
-	return b.store.ListApplicationVersions(applicationId)
+func (b *Backend) ListApplicationVersions(applicationId string) ([]*types.ApplicationVersion, error) {
+	return b.store.GetVersions(applicationId)
```
Comment (cmingxu) [inline backend/list.go]: does GetVersionsByApplicationId more suitable?

File: backend/rollback.go [hunk 4]
```diff
@@ -14,35 +13,27 @@ import (
...
-	}
-
 	// Update application status to ROLLINGBACK
-	if err := b.store.UpdateApplicationStatus(app.ID, "ROLLINGBACK"); err != nil {
+	if err := b.store.UpdateAppStatus(app.ID, "ROLLINGBACK"); err != nil {
 		return err
 	}
 
-	versions, err := b.store.ListApplicationVersions(applicationId)
+	versions, err := b.store.GetAndSortVersions(applicationId)
```
Comment (cmingxu) [inline backend/rollback.go]: GetAndSortVersions => GetSortedVersions

Comment (cmingxu) [inline backend/rollback.go]: Or

GetVersions(Criteria: AppId, SortBy: Whatever)

File: backend/rollback.go [hunk 5]
```diff
@@ -14,35 +13,27 @@ import (
...
 	}
 
-	sort.Strings(versions)
-
-	rollbackVer := versions[len(versions)-2]
-	version, err := b.store.FetchApplicationVersion(applicationId, rollbackVer)
-	if err != nil {
-		return err
+	if len(versions) < 2 {
+		return errors.New("not have history version to rollback")
```
Comment (cmingxu) [inline backend/rollback.go]: "not have history version to rollback"

=> 
no version to rollback

File: backend/rollback.go [hunk 6]
```diff
@@ -14,35 +13,27 @@ import (
...
 	}
+	version := versions[len(versions)-2]
```
Comment (cmingxu) [inline backend/rollback.go]: How about define a struct represent group of versions?
Have a methods VersionGroup.CurrentVersion VersionGroup.LastVersion VersionGroup.LastHealthVersion

File: backend/rollback.go [hunk 7]
```diff
@@ -14,35 +13,27 @@ import (
...
 
-	tasks, err := b.store.ListApplicationTasks(applicationId)
+	tasks, err := b.store.GetTasks(applicationId)
```
Comment (cmingxu) [inline backend/rollback.go]: Confused about 
 b.store.GetTasks(applicationId)
 b.store.GetTasks(id)

File: backend/rollback.go [hunk 8]
```diff
@@ -99,37 +90,32 @@ func (b *Backend) RollbackApplication(applicationId string) error {
...
-					Address:  *task.AgentHostname,
-					Port:     int(*taskInfo.Container.Docker.PortMappings[0].HostPort),
+					Address:  task.AgentHostname,
+					Port:     int64(*taskInfo.Container.Docker.PortMappings[0].HostPort),
 					TaskID:   task.Name,
 					AppID:    app.ID,
 					Protocol: healthCheck.Protocol,
-					Interval: int(healthCheck.IntervalSeconds),
-					Timeout:  int(healthCheck.TimeoutSeconds),
+					Interval: int64(healthCheck.IntervalSeconds),
```
Comment (cmingxu) [inline backend/rollback.go]: is it OK use time.Duration as Interval & Timeout?

File: backend/rollback.go [hunk 9]
```diff
@@ -99,37 +90,32 @@ func (b *Backend) RollbackApplication(applicationId string) error {
...
 
-				if healthCheck.Path != nil {
-					check.Path = *healthCheck.Path
-				}
-
-				if healthCheck.MaxConsecutiveFailures != nil {
-					check.MaxFailures = *healthCheck.MaxConsecutiveFailures
-				}
+				check.Path = healthCheck.Path
+				check.MaxFailures = healthCheck.MaxConsecutiveFailures
```
Comment (cmingxu) [inline backend/rollback.go]: MaxConsecutiveFailures confused me, does the non-consecutive failures exists?

File: backend/rollback.go [hunk 10]
```diff
@@ -138,7 +124,7 @@ func (b *Backend) RollbackApplication(applicationId string) error {
 		b.sched.Status = "idle"
 	}
 
-	if err := b.store.UpdateApplicationStatus(app.ID, "RUNNING"); err != nil {
+	if err := b.store.UpdateAppStatus(app.ID, "RUNNING"); err != nil {
```
Comment (cmingxu) [inline backend/rollback.go]: I saw "RUNNING" strings scatter around, do we plan to constantnize them

File: backend/update.go [hunk 11]
```diff
@@ -30,18 +31,62 @@ func (b *Backend) UpdateApplication(applicationId string, instances int, version
...
+		taskNames = append(taskNames, task.Name)
+	}
+
+	go func() {
+		for {
+			select {
+			case event := <-b.sched.GetEvent(sched.Event_UPDATE):
+				status := event.GetUpdate().GetStatus()
+				ID := status.TaskId.GetValue()
+				taskName := strings.Split(ID, "-")[1]
```
Comment (cmingxu) [inline backend/update.go]: how about make this a util function

File: backend/update.go [hunk 12]
```diff
@@ -30,18 +31,62 @@ func (b *Backend) UpdateApplication(applicationId string, instances int, version
...
+				for _, name := range taskNames {
+					if name == taskName {
+						switch status.GetState() {
+						case mesos.TaskState_TASK_FAILED,
+							mesos.TaskState_TASK_FINISHED,
+							mesos.TaskState_TASK_LOST:
+							maxFailover++
+							if maxFailover >= int(version.UpdatePolicy.MaxFailovers) {
+								switch version.UpdatePolicy.Action {
```
Comment (cmingxu) [inline backend/update.go]: switch strings.ToLower() ?

File: backend/update.go [hunk 13]
```diff
@@ -30,18 +31,62 @@ func (b *Backend) UpdateApplication(applicationId string, instances int, version
...
+				ID := status.TaskId.GetValue()
+				taskName := strings.Split(ID, "-")[1]
+				for _, name := range taskNames {
+					if name == taskName {
+						switch status.GetState() {
+						case mesos.TaskState_TASK_FAILED,
+							mesos.TaskState_TASK_FINISHED,
+							mesos.TaskState_TASK_LOST:
+							maxFailover++
+							if maxFailover >= int(version.UpdatePolicy.MaxFailovers) {
```
Comment (cmingxu) [inline backend/update.go]: Kind of nested too much

File: backend/update.go [hunk 14]
```diff
@@ -112,62 +157,56 @@ func (b *Backend) UpdateApplication(applicationId string, instances int, version
...
-					app, err := b.store.FetchApplication(applicationId)
+					app, err := b.store.GetApp(applicationId)
 					if err != nil {
 						return err
 					}
 
 					// Rest application updated instance count to zero.
 					if app.UpdatedInstances == app.Instances {
-						if err := b.store.ResetApplicationUpdatedInstances(app.ID); err != nil {
+						if err := b.store.UpdateAppUpdatedInstance(app.ID, 0); err != nil {
```
Comment (cmingxu) [inline backend/update.go]: UpdateAppUpdatedInstance => UpdateAppScaleInstance ?

File: main.go [hunk 15]
```diff
@@ -56,15 +58,17 @@ func main() {
 		FailoverTimeout: proto.Float64(60 * 60 * 24 * 7),
 	}
 
-	store, err := consul.NewConsul(*consulAddr)
+	db, err := bolt.Open("./bin/bolt.db", 0644, nil)
```
Comment (cmingxu) [inline main.go]: make a common place to store all data. /var/lib/swan ?

File: store/boltdb/app.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+			return bkt.Put(bucketKeyData, p)
+		}); err != nil {
+			return err
+		}
+	}
+
+	return tx.Commit()
+}
+
+func (db *Boltdb) GetApp(appId string) (*types.Application, error) {
```
Comment (cmingxu) [inline store/boltdb/app.go]: GetApp => GetAppById ?

File: store/boltdb/app.go [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	if len(apps) < 1 {
+		return nil, errAppUnknown
+	}
+
+	return apps[0], nil
+}
+
+func (db *Boltdb) GetApps(appIds ...string) ([]*types.Application, error) {
```
Comment (cmingxu) [inline store/boltdb/app.go]: How to list apps with certain conditions?

File: store/boltdb/app.go (hunk 16 above)
Comment (upccup) [inline store/boltdb/app.go]: we can get app only by id

**Comment by xiaods** [comment]
PR 太多。我无法接受这样的 PR

**Pull request closed without merging**